File: python/enolib/elements/section.py
from ..errors.validation import Validation
from . import element as element_module
from . import embed
from . import field
from . import flag
from .element_base import ElementBase
from .missing import missing_embed
from .missing import missing_field
from .missing import missing_flag
from .missing import missing_section
from .missing import missing_section_element
from ..constants import (
    DOCUMENT,
    EMBED_BEGIN,
    FIELD,
    FLAG,
    SECTION
)

class Section(ElementBase):
    def __init__(self, context, instruction, parent=None):
        super().__init__(context, instruction, parent)

        self._all_elements_required = parent._all_elements_required if parent else False

    def __repr__(self):
        if self._instruction['type'] == DOCUMENT:
            return f"<class Section document elements={len(self._elements())}>"

        return f"<class Section key=\"{self._instruction['key']}\" elements={len(self._elements())}>"

    def _element(self, key, *, required=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        if len(elements) == 0:
            if required or required is None and self._all_elements_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_element')
            elif required is None:
                return missing_section_element.MissingSectionElement(key, self)
            else:
                return None

        if len(elements) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [element._instruction for element in elements],
                'expected_single_element'
            )

        return elements[0]

    def _elements(self, as_map=False):
        if not hasattr(self, '_instantiated_elements'):
            self._instantiated_elements_map = {}
            
            def instantiate_and_index(element):
                if element['type'] == EMBED_BEGIN:
                    instance = embed.Embed(self._context, element, self)
                elif element['type'] == FIELD:
                    instance = field.Field(self._context, element, self)
                elif element['type'] == FLAG:
                    instance = flag.Flag(self._context, element, self)
                elif element['type'] == SECTION:
                    instance = Section(self._context, element, self)

                if element['key'] in self._instantiated_elements_map:
                    self._instantiated_elements_map[element['key']].append(instance)
                else:
                    self._instantiated_elements_map[element['key']] = [instance]

                return instance

            self._instantiated_elements = [instantiate_and_index(element) for element in self._instruction['elements']]

        return self._instantiated_elements_map if as_map else self._instantiated_elements

    def _embed(self, key, *, required=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        if len(elements) == 0:
            if required or required is None and self._all_elements_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_embed')
            elif required is None:
                return missing_embed.MissingEmbed(key, self)
            else:
                return None

        if len(elements) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [element._instruction for element in elements],
                'expected_single_embed'
            )

        element = elements[0]

        if element._instruction['type'] != EMBED_BEGIN:
            raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_embed')

        return element

    def _flag(self, key, *, required=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        if len(elements) == 0:
            if required or required is None and self._all_elements_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_flag')
            elif required is None:
                return missing_flag.MissingFlag(key, self)
            else:
                return None

        if len(elements) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [element._instruction for element in elements],
                'expected_single_flag'
            )

        element = elements[0]

        if element._instruction['type'] != FLAG:
            raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_flag')

        return element

    def _field(self, key, *, required=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        if len(elements) == 0:
            if required or required is None and self._all_elements_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_field')
            elif required is None:
                return missing_field.MissingField(key, self)
            else:
                return None

        if len(elements) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [element._instruction for element in elements],
                'expected_single_field'
            )

        element = elements[0]

        if element._instruction['type'] != FIELD:
            raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_field')

        return element

    def _missing_error(self, element):
        if isinstance(element, missing_field.MissingField):
            raise Validation.missing_element(self._context, element._key, self._instruction, 'missing_field')
        elif isinstance(element, missing_section.MissingSection):
            raise Validation.missing_element(self._context, element._key, self._instruction, 'missing_section')
        else:
            raise Validation.missing_element(self._context, element._key, self._instruction, 'missing_element')

    def _section(self, key, *, required=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        if len(elements) == 0:
            if required or required is None and self._all_elements_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_section')
            elif required is None:
                return missing_section.MissingSection(key, self)
            else:
                return None

        if len(elements) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [element._instruction for element in elements],
                'expected_single_section'
            )

        element = elements[0]

        if element._instruction['type'] != SECTION:
            raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_section')

        return element

    def _untouched(self):
        if not hasattr(self, '_touched'):
            return self._instruction

        for element in self._elements():
            untouched_element = element._untouched()
            if untouched_element:
                return untouched_element

        return False

    def all_elements_required(self, required=True):
        self._all_elements_required = required

        for element in self._elements():
            if element._instruction['type'] == SECTION:
                element.all_elements_required(required)
            elif element._instruction['type'] == FIELD:
                element.all_attributes_required(required)

    def assert_all_touched(self, message=None, *, only=None, skip=None):
        elements_map = self._elements(True)

        for key, elements in elements_map.items():
            if (skip and key in skip) or (only and key not in only):
                continue

            for element in elements:
                untouched = element._untouched()

                if untouched:
                    if callable(message):
                        message = message(element_module.Element(self._context, untouched, self))

                    raise Validation.unexpected_element(self._context, message, untouched)

    def element(self, key=None):
        return self._element(key)

    def elements(self, key=None):
        self._touched = True

        if key:
            elements_map = self._elements(True)
            return elements_map[key] if key in elements_map else []

        return self._elements()

    def embed(self, key=None):
        return self._embed(key)

    def embeds(self, key=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        def typecheck(element):
            if element._instruction['type'] != EMBED_BEGIN:
                raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_embeds')

            return element

        return [typecheck(element) for element in elements]
        
    def field(self, key=None):
        return self._field(key)

    def fields(self, key=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        def typecheck(element):
            if element._instruction['type'] != FIELD:
                raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_fields')

            return element

        return [typecheck(element) for element in elements]

    def flag(self, key=None):
        return self._flag(key)

    def flags(self, key=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        def typecheck(element):
            if element._instruction['type'] != FLAG:
                raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_flags')

            return element

        return [typecheck(element) for element in elements]

    def optional_element(self, key):
        return self._element(key, required=False)

    def optional_embed(self, key):
        return self._embed(key, required=False)

    def optional_field(self, key):
        return self._field(key, required=False)

    def optional_flag(self, key):
        return self._flag(key, required=False)

    def optional_section(self, key):
        return self._section(key, required=False)

    def parent(self):
        if self._instruction['type'] == DOCUMENT:
            return None

        return self._parent or Section(self._context, self._instruction['parent'])

    def required_element(self, key=None):
        return self._element(key, required=True)

    def required_embed(self, key=None):
        return self._embed(key, required=True)

    def required_field(self, key=None):
        return self._field(key, required=True)

    def required_flag(self, key=None):
        return self._flag(key, required=True)

    def required_section(self, key=None):
        return self._section(key, required=True)

    def section(self, key=None):
        return self._section(key)

    def sections(self, key=None):
        self._touched = True

        if not key:
            elements = self._elements()
        else:
            elements_map = self._elements(True)
            elements = elements_map[key] if key in elements_map else []

        def typecheck(element):
            if element._instruction['type'] != SECTION:
                raise Validation.unexpected_element_type(self._context, key, element._instruction, 'expected_sections')

            return element

        return [typecheck(element) for element in elements]

    def touch(self):
        self._touched = True

        for element in self._elements():
            element.touch()

File: ruby/lib/enolib/elements/section.rb
# frozen_string_literal: true

module Enolib
  class Section < ElementBase
    attr_reader :instruction
    
    def initialize(context, instruction, parent = nil)
      super(context, instruction, parent)

      @all_elements_required = parent ? parent.all_elements_required? : false
    end

    def _missing_error(element)
      case element
      when MissingEmbed
        raise Errors::Validation.missing_element(@context, element.instance_variable_get(:@key), @instruction, 'missing_embed')
      when MissingField
        raise Errors::Validation.missing_element(@context, element.instance_variable_get(:@key), @instruction, 'missing_field')
      when MissingFlag
        raise Errors::Validation.missing_element(@context, element.instance_variable_get(:@key), @instruction, 'missing_flag')
      when MissingSection
        raise Errors::Validation.missing_element(@context, element.instance_variable_get(:@key), @instruction, 'missing_section')
      else
        raise Errors::Validation.missing_element(@context, element.instance_variable_get(:@key), @instruction, 'missing_element')
      end
    end

    def _untouched
      return @instruction unless instance_variable_defined?(:@touched)

      _elements.each do |element|
        untouched_element = element._untouched
        return untouched_element if untouched_element
      end

      false
    end

    def all_elements_required(required = true)
      @all_elements_required = required

      _elements.each do |element|
        if element.instruction[:type] == :section
          element.all_elements_required(required)
        elsif element.instruction[:type] == :field
          element.all_attributes_required(required)
        end
      end
    end

    def all_elements_required?
      @all_elements_required
    end

    def assert_all_touched(message = nil, except: nil, only: nil, &block)
      message = Proc.new(&block) if block_given?

      _elements(map: true).each do |key, elements|
        next if except && except.include?(key) || only && !only.include?(key)

        elements.each do |element|
          untouched = element._untouched

          next unless untouched

          if message.is_a?(Proc)
            message = message.call(Element.new(@context, untouched, self))
          end

          raise Errors::Validation.unexpected_element(@context, message, untouched)
        end
      end
    end

    def element(key = nil)
      _element(key)
    end

    def elements(key = nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements_map.has_key?(key) ? elements_map[key] : []
      else
        _elements
      end
    end
    
    def embed(key = nil)
      _embed(key)
    end
    
    def embeds(key = nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      elements.each do |element|
        next if element.is_a?(Embed)
        raise Errors::Validation.unexpected_element_type(@context, key, element.instruction, 'expected_embeds')
      end
      
      elements
    end

    def field(key = nil)
      _field(key)
    end

    def fields(key = nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      elements.each do |element|
        next if element.is_a?(Field)
        raise Errors::Validation.unexpected_element_type(@context, key, element.instruction, 'expected_fields')
      end
      
      elements
    end
    
    def flag(key = nil)
      _flag(key)
    end
    
    def flags(key = nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      elements.each do |element|
        next if element.is_a?(Flag)
        raise Errors::Validation.unexpected_element_type(@context, key, element.instruction, 'expected_flags')
      end
      
      elements
    end

    def optional_element(key = nil)
      _element(key, required: false)
    end

    def optional_embed(key = nil)
      _embed(key, required: false)
    end

    def optional_field(key = nil)
      _field(key, required: false)
    end

    def optional_flag(key = nil)
      _flag(key, required: false)
    end

    def optional_section(key = nil)
      _section(key, required: false)
    end

    def parent
      if @instruction[:type] == :document
        nil
      else
        @parent || Section.new(@context, @instruction[:parent])
      end
    end

    def required_element(key = nil)
      _element(key, required: true)
    end

    def required_embed(key = nil)
      _embed(key, required: true)
    end

    def required_field(key = nil)
      _field(key, required: true)
    end
    
    def required_flag(key = nil)
      _flag(key, required: true)
    end

    def required_section(key = nil)
      _section(key, required: true)
    end

    def section(key = nil)
      _section(key)
    end

    def sections(key = nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      elements.each do |element|
        next if element.is_a?(Section)
        raise Errors::Validation.unexpected_element_type(@context, key, element.instruction, 'expected_sections')
      end
      
      elements
    end

    def to_s
      if @instruction[:type] == :document
        "#<Enolib::Section document elements=#{elements.length}>"
      else
        "#<Enolib::Section key=#{@instruction[:key]} elements=#{elements.length}>"
      end
    end

    def touch
      @touched = true

      _elements.each(&:touch)
    end

    private

    def _element(key = nil, required: nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      if elements.empty?
        if required || required == nil && @all_elements_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_element')
        elsif required == nil
          return MissingSectionElement.new(key, self)
        else
          return nil
        end
      end

      if elements.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context, key, elements.map(&:instruction), 'expected_single_element'
        )
      end

      elements[0]
    end

    def _elements(map: false)
      unless instance_variable_defined?(:@instantiated_elements)
        @instantiated_elements_map = {}
        @instantiated_elements = @instruction[:elements].map do |element|
          instance =
            case element[:type]
            when :embed_begin then Embed.new(@context, element, self)
            when :field then Field.new(@context, element, self)
            when :flag then Flag.new(@context, element, self)
            when :section then Section.new(@context, element, self)
            end

          if @instantiated_elements_map.has_key?(element[:key])
            @instantiated_elements_map[element[:key]].push(instance)
          else
            @instantiated_elements_map[element[:key]] = [instance]
          end

          instance
        end
      end

      map ? @instantiated_elements_map : @instantiated_elements
    end

    def _embed(key = nil, required: nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      if elements.empty?
        if required || required == nil && @all_elements_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_embed')
        elsif required == nil
          return MissingEmbed.new(key, self)
        else
          return nil
        end
      end

      if elements.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context, key, elements.map(&:instruction), 'expected_single_embed'
        )
      end

      element = elements[0]

      unless element.is_a?(Embed)
        raise Errors::Validation.unexpected_element_type(
          @context, key, element.instruction, 'expected_embed'
        )
      end

      element
    end

    def _field(key = nil, required: nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      if elements.empty?
        if required || required == nil && @all_elements_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_field')
        elsif required == nil
          return MissingField.new(key, self)
        else
          return nil
        end
      end

      if elements.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context, key, elements.map(&:instruction), 'expected_single_field'
        )
      end

      element = elements[0]

      unless element.is_a?(Field)
        raise Errors::Validation.unexpected_element_type(
          @context, key, element.instruction, 'expected_field'
        )
      end

      element
    end
    
    def _flag(key = nil, required: nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      if elements.empty?
        if required || required == nil && @all_elements_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_flag')
        elsif required == nil
          return MissingFlag.new(key, self)
        else
          return nil
        end
      end

      if elements.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context, key, elements.map(&:instruction), 'expected_single_flag'
        )
      end

      element = elements[0]

      unless element.is_a?(Flag)
        raise Errors::Validation.unexpected_element_type(
          @context, key, element.instruction, 'expected_flag'
        )
      end

      element
    end

    def _section(key = nil, required: nil)
      @touched = true

      if key
        elements_map = _elements(map: true)
        elements = elements_map.has_key?(key) ? elements_map[key] : []
      else
        elements = _elements
      end

      if elements.empty?
        if required || required == nil && @all_elements_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_section')
        elsif required == nil
          return MissingSection.new(key, self)
        else
          return nil
        end
      end

      if elements.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context, key, elements.map(&:instruction), 'expected_single_section'
        )
      end

      element = elements[0]

      unless element.is_a?(Section)
        raise Errors::Validation.unexpected_element_type(
          @context, key, element.instruction, 'expected_section'
        )
      end

      element
    end
  end
end
